### PR TITLE
Add error fields for cluster and refactor machine status

### PR DESCRIPTION
**Pull request by maelk**
- add validation of cluster spec
- add errorReason and errorMessage to cluster Status
- refactor machine status

### CODE REVIEW

**Comment by maelk** [comment]
@xenwar

File: config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml [hunk 1]
```diff
@@ -132,83 +132,41 @@ spec:
...
+                a more verbose string suitable for logging and human consumption.
+                \n This field should not be set for transitive errors that a controller
+                faces that are expected to be fixed automatically over time (like
+                service outages), but instead indicate that something is fundamentally
+                wrong with the BaremetalMachine's spec or the configuration of the
+                controller, and that manual intervention is required. Examples of
+                terminal errors would be invalid combinations of settings in the spec,
+                values that are unsupported by the controller, or the responsible
+                controller itself being critically misconfigured. \n Any transient
+                errors that occur during the reconciliation of Machines can be added
```
Comment (kashifest) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml]: BaremetalMachine or Machine??

File: config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml [hunk 2]
```diff
@@ -132,83 +132,41 @@ spec:
...
+                problem reconciling the BaremetalMachine and will contain a succinct
+                value suitable for machine interpretation. \n This field should not
+                be set for transitive errors that a controller faces that are expected
+                to be fixed automatically over time (like service outages), but instead
+                indicate that something is fundamentally wrong with the BaremetalMachine's
                 spec or the configuration of the controller, and that manual intervention
                 is required. Examples of terminal errors would be invalid combinations
                 of settings in the spec, values that are unsupported by the controller,
                 or the responsible controller itself being critically misconfigured.
                 \n Any transient errors that occur during the reconciliation of Machines
```
Comment (kashifest) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml]: BaremetalMachine or Machine??

File: config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml [hunk 3]
```diff
@@ -132,83 +132,41 @@ spec:
...
-                can be added as events to the Machine object and/or logged in the
-                controller's output."
+                can be added as events to the BaremetalMachine object and/or logged
+                in the controller's output."
               type: string
-            infrastructureReady:
```
Comment (kashifest) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml]: Why are we removing this?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml (hunk 1 above)
Comment (maelk) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml]: fixed, thanks

File: config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml (hunk 2 above)
Comment (maelk) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_baremetalmachines.yaml]: fixed, thanks

**Pull request merged**